Mage.Sets/src/mage/cards/m/MagmaPummeler.java

```diff
@@ -50,40 +52,76 @@ public MagmaPummeler copy() {
     }
 }
 
-class MagmaPummelerEffect extends PreventDamageAndRemoveCountersEffect {
+class MagmaPummelerPreventionEffect extends PreventDamageAndRemoveCountersEffect {
 
-    MagmaPummelerEffect() {
+    MagmaPummelerPreventionEffect() {
         super(true, true, true);
-        staticText += ". When one or more counters are removed from {this} this way, it deals that much damage to any target";
     }
 
-    private MagmaPummelerEffect(final MagmaPummelerEffect effect) {
+    private MagmaPummelerPreventionEffect(final MagmaPummelerPreventionEffect effect) {
         super(effect);
     }
 
     @Override
-    public MagmaPummelerEffect copy() {
-        return new MagmaPummelerEffect(this);
+    public PreventDamageAndRemoveCountersEffect copy() {
+        return new MagmaPummelerPreventionEffect(this);
     }
 
     @Override
-    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
-        Permanent permanent = game.getPermanent(source.getSourceId());
-        if (permanent == null) {
+    protected void additionalEffect(GameEvent event, Ability source, Game game, int amountRemovedInTotal, int amountRemovedThisTime) {
+        super.additionalEffect(event, source, game, amountRemovedInTotal, amountRemovedThisTime);
+        // This is not the most elegant, but ensure we only trigger at most once per DAMAGED_BATCH_COULD_HAVE_FIRED,
+        // with the correct number of counters removed (the last time we stored value it will be the total)
+        source.getSubAbilities()
+                .stream()
+                .filter(MagmaPummelerTriggeredAbility.class::isInstance)
+                .forEach(ability -> ((MagmaPummelerTriggeredAbility) ability).setStoredValue(amountRemovedInTotal));
+    }
+}
+
+/**
+ * Having a trigger from a replacement effect is a weird setup for a trigger.
+ * We manage with the DAMAGED_BATCH_COULD_HAVE_FIRED special event, and the stored
+ * amount in {@link mage.abilities.dynamicvalue.common.SavedCounterRemovedValue}.
+ */
+class MagmaPummelerTriggeredAbility extends TriggeredAbilityImpl {
```

Comment:
Old card version did fire 1 reflexive trigger per source of damage prevented. Current only fires a single trigger.

Reply:
I mean reflexive trigger and normal trigger are different due mtg rules. Current code uses wrong trigger type (normal), but must be a reflexive:

> If an amount of damage would be dealt to Magma Pummeler greater than the number of +1/+1 counters on it, all of that damage is prevented, and all of those counters are removed. In most cases, this will result in Magma Pummeler having 0 toughness, so it will be put into its owner's graveyard. Its reflexive triggered ability still triggers and it will still deal that much damage to any target.
(2021-11-19)